src/sources/util/unix_stream.rs

```diff
@@ -34,15 +35,35 @@ use crate::{
 /// syslog source).
 pub fn build_unix_stream_source(
     listen_path: PathBuf,
+    socket_file_mode: Option<u32>,
     decoder: Decoder,
     handle_events: impl Fn(&mut [Event], Option<Bytes>) + Clone + Send + Sync + 'static,
     shutdown: ShutdownSignal,
     out: SourceSender,
-) -> Source {
-    Box::pin(async move {
-        let listener = UnixListener::bind(&listen_path).expect("Failed to bind to listener socket");
-        info!(message = "Listening.", path = ?listen_path, r#type = "unix");
+) -> crate::Result<Source> {
+    let listener = UnixListener::bind(&listen_path).expect("Failed to bind to listener socket");
+    info!(message = "Listening.", path = ?listen_path, r#type = "unix");
+
+    if let Some(socket_permission) = socket_file_mode {
+        match fs::set_permissions(&listen_path, fs::Permissions::from_mode(socket_permission)) {
+            Ok(_) => info!(
+                message = "Socket permissions updated.",
+                permission = socket_permission
+            ),
+            Err(e) => {
+                if let Err(error) = remove_file(&listen_path) {
+                    emit!(UnixSocketFileDeleteError {
+                        path: &listen_path,
+                        error
+                    });
+                }
+
+                return Err(Box::new(e));
+            },
+        };
+    };
```

Comment:
Could this be moved into a helper function to reduce duplication?

Reply:
What I had before but finally reverted haha, will do!